packages/statemanager/src/stateManager.ts

```diff
@@ -157,43 +318,41 @@ export class DefaultStateManager extends BaseStateManager implements StateManage
    */
   async getContractCode(address: Address): Promise<Uint8Array> {
     const account = await this.getAccount(address)
+    if (!account) {
+      return new Uint8Array(0)
+    }
     if (!account.isContract()) {
       return new Uint8Array(0)
     }
     const key = this._prefixCodeHashes
       ? concatBytes(CODEHASH_PREFIX, account.codeHash)
       : account.codeHash
-    // @ts-expect-error
-    const code = await this._trie._db.get(key)
-    return code ?? new Uint8Array(0)
-  }
 
-  /**
-   * Creates a storage trie from the primary storage trie
-   * for an account and saves this in the storage cache.
-   * @private
-   */
-  async _lookupStorageTrie(address: Address): Promise<Trie> {
-    // from state trie
-    const account = await this.getAccount(address)
-    const storageTrie = this._trie.copy(false)
-    storageTrie.root(account.storageRoot)
-    storageTrie.flushCheckpoints()
-    return storageTrie
+    const keyHex = bytesToHex(key)
+    if (keyHex in this._codeCache) {
+      return this._codeCache[keyHex]
+    } else {
+      // @ts-expect-error
+      const code = (await this._trie._db.get(key)) ?? new Uint8Array(0)
+      this._codeCache[keyHex] = code
```

Comment:
Had to take a look, but it seems like we do not persist code cache / storage trie cache if we set state root. So this code/storage tries cache only works within a block. Beyond the scope of this PR but we could also (possibly re-use the account cache class?) to do the same for codeCache/storage trie cache?

Reply:
Yes, definitely. 🙂 Already thought about doing this but refrained since things are complex enough. But we can and likely want to add this third cache piece as well I guess.